repositories/EventRepository.ts

```diff
@@ -33,16 +33,51 @@ export class EventRepository extends BaseRepository<EventModel> {
   }
 
   public async findByAttendanceCode(attendanceCode: string): Promise<EventModel> {
-    return this.repository.findOne({ attendanceCode });
+    const matchingEvents = await this.repository.find({ attendanceCode });
+
+    // Find all events with the given attendance code
+    const validEvents = matchingEvents.filter((event) => !event.isTooEarlyToAttendEvent() && !event.isTooLateToAttendEvent());
+
+    // If there are eligible events, return the first one
+    if (validEvents.length > 0) {
+      return validEvents[0];
+    }
+
+    // Otherwise, find the closest event to the current time
+    const currentTime = new Date();
+    let closestEvent = null;
+    let closestTimeDifference = Infinity;
+
+    matchingEvents.forEach((event) => {
+      const eventStartTime = new Date(event.start);
+      const timeDifference = Math.abs(eventStartTime.getTime() - currentTime.getTime());
+
+      // Update closest event if necessary
+      if (timeDifference < closestTimeDifference) {
+        closestEvent = event;
+        closestTimeDifference = timeDifference;
+      }
+    });
+
+    return closestEvent;
```

Comment:
Just curious, what's the edge case that warrants this code? Shouldn't all potentially valid events be caught in the above code? I know we're validating it one more time in the AttendanceService, just curious what the reasoning for adding it was

Reply:
And side note, if there is a use case that warrants this code, does it make more sense to use validEvents instead of matchingEvents?

Reply:
The use case is the small pop-up that announces whether an event "hasn't occurred yet / has already occurred / doesn't exist" when the user enters an attendance code that isn't matched to an event currently going on, in AttendanceService's validateEventToAttend method.

 The way it used to work when attendance codes were unique is that the validate method would simply check if the one event with the matching attendance code hasn't started, has already ended, or doesn't exist in the schema. 

However with repeated attendance codes, it is possible for events sharing the same code to be simultaneously "too early" AND "too late," so I chose to return the closest event to the current date as a "guess" as to what the user meant to check in to when they entered a code. For example, if code "ABC" was used both for an event a year ago and two days in the future, I'm assuming that the user meant to check in to the one two days in the future, and will show the "hasn't occurred yet" message. Let me know if this implementation needs more consideration.

Reply:
I believe we shouldn't have the "hasn't occurred yet" message at all as it give away the checkin code for a future event, nullifying the secrecy of the checkin code.